File: translated/f-to-c/DRB164-simdmissinglock1-orig-gpu-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Concurrent access of var@30:13 without acquiring locks causes atomicity violation. Data race present.
//Data Race Pairs, var@30:13:W vs. var@30:22:R

#include <omp.h>
#include <stdio.h>

// Global variables (from module)
int var[16];
int i, j;

int main()
{
    for (i = 0; i < 16; i++) {
        var[i] = 0;
    }

    #pragma omp target map(tofrom:var[0:16]) device(0)
    #pragma omp teams distribute parallel for
    for (i = 0; i < 20; i++) {
        #pragma omp simd
        for (j = 0; j < 16; j++) {
            var[j] = var[j] + 1;  // Data race: no lock or reduction protection
        }
    }

    printf("%d\n", var[15]);  // Adjust for 0-based indexing

    return 0;
} 

File: translated/f-to-go/DRB164-simdmissinglock1-orig-gpu-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Concurrent access of var@30:13 without acquiring locks causes atomicity violation. Data race present.
//Data Race Pairs, var@30:13:W vs. var@30:22:R

package main

import (
	"fmt"
	"sync"
)

// Package-level variables (module equivalent)
var variable [16]int

func main() {
	// Initialize
	for i := 1; i <= 16; i++ {
		variable[i-1] = 0
	}

	//$omp target map(tofrom:var) device(0)
	//$omp teams distribute parallel do
	// MISSING: reduction clause
	var wg sync.WaitGroup

	for i := 1; i <= 20; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			
			//$omp simd
			for j := 1; j <= 16; j++ {
				variable[j-1] = variable[j-1] + 1 // RACE: Multiple threads + SIMD accessing shared variable
			}
			//$omp end simd
		}()
	}
	wg.Wait()
	//$omp end teams distribute parallel do
	//$omp end target

	fmt.Printf("%d\n", variable[15])
}